File: go/src/shapes.go
package main

import "math"

type Shape interface {
	Area() float64
}

type Rectangle struct {
	height float64
	width  float64
}

type Circle struct {
	radius float64
}

type Triangle struct {
	base   float64
	height float64
}

func Perimeter(rect Rectangle) float64 {
	return 2 * (rect.height + rect.width)
}

func (r Rectangle) Area() float64 {
	return r.height * r.width
}

func (c Circle) Area() float64 {
	return math.Pi * c.radius * c.radius
}

func (t Triangle) Area() float64 {
	return 0.5 * t.base * t.height
}

File: rust/src/shapes.rs
use std::f64::consts::PI;

trait HasArea {
    fn area(&self) -> f64;
}

struct Rectangle {
    height: f64,
    weight: f64,
}

struct Circle {
    radius: f64,
}

struct Triangle {
    base: f64,
    height: f64,
}

impl HasArea for Rectangle {
    fn area(&self) -> f64 {
        self.height * self.weight
    }
}

impl HasArea for Circle {
    fn area(&self) -> f64 {
        PI * &self.radius * &self.radius
    }
}

impl HasArea for Triangle {
    fn area(&self) -> f64 {
        0.5 * &self.height * self.base
    }
}

fn get_area<T: HasArea>(shape: T) -> f64 {
    return shape.area();
}

#[cfg(test)]
mod tests {
    use std::f64::consts::PI;

    use crate::shapes::{Circle, get_area, Rectangle, Triangle};

    #[test]
    fn test_get_area() {
        assert_eq!(get_area(Rectangle { height: 12.0, weight: 6.0 }), 72.0);
        assert_eq!(get_area(Circle { radius: 10.0 }), PI * 100.0);
        assert_eq!(get_area(Triangle { base: 12.0, height: 6.0 }), 36.0);
    }
}